### PR TITLE
Recipe to format SQL statements inside text blocks

### PR DESCRIPTION
**Pull request by MahatmaFatalError**
Successor of https://github.com/openrewrite/rewrite-migrate-java/pull/220

### CODE REVIEW

**Comment by timtebeek** [comment]
@MahatmaFatalError I've just pushed a small change to get your recipe to run; that ought to help you work towards fixing the recipe and tests over there. :)

**Comment by MahatmaFatalError** [comment]
Thanks for your support. It seems the indentation is wrong. Do you know if there is a reusable component for text block indentation, maybe related to https://github.com/MahatmaFatalError/rewrite-migrate-java/blob/cedac29a8036abb6f5e69037e3088ffd394f1d8f/src/main/java/org/openrewrite/java/migrate/lang/UseTextBlocks.java#LL309C27-L309C37 ? Otherwise I would copy the code

**Comment by timtebeek** [comment]
That I don't know... @kunli2 might know if it makes sense to open up some of that functionality to other recipes such as this one. Until then we can start with a copy to get this PR going.

**Comment by MahatmaFatalError** [comment]
Alright, tried to fix the indentation problem, but the test fails with `Expected recipe to complete in 1 cycle, but took 2 cycles.` I don't know what to do here

**Comment by timtebeek** [comment]
Not quite sure about a fix for the cycles; asking a colleague to have look since I have to rush out.

**Comment by shanman190** [comment]
So generally it's often better to just call `JavaVisitor#autoFormat(J, ExecutionContext, Cursor)` which will automatically use the detected style of the insertion point so that you don't have to do anything around finding the proper indentation type or level. It should take care of everything for you.

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+                        // add last line to ensure the closing delimiter is in a new line to manage
+                        // indentation & remove theformatted
+                        // need to escape ending quote in the content
+                        boolean isEndsWithNewLine = literal.getValueSource().endsWith("\n");
+                        if (!isEndsWithNewLine) {
+                            formatted = formatted + "\\\n" + indentation;
+                        }
+
+                        return literal.withValueSource(String.format("\"\"\"%s\"\"\"", formatted));
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/sql/FormatSql.java]: This is basically what causes the test to fail. The recipe "modifies" the LST on a second cycle with a no-op modification. As checking for modifications using `Object#equals(Object)` would be too expensive, OpenRewrite uses referential equality. This in turn means that recipes should try hard to only modify the LST when necessary.

```suggestion
                        String newValueSource = String.format("\"\"\"%s\"\"\"", formatted);
                        if ( !newValueSource.equals(literal.getValueSource())) {
                            return literal.withValueSource(newValueSource);
                        }
```

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import com.github.vertical_blank.sqlformatter.SqlFormatter;
+import com.github.vertical_blank.sqlformatter.languages.Dialect;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class FormatSql extends Recipe {
+    @Option(displayName = "SQL dialect to be used to format SQL snippets.", description = "", example = "PostgreSql", required = false)
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/sql/FormatSql.java:44]: Here we should enumerate all allowed values using `Option#valid`. E.g.
```java
@Option(valid = {"foo", "bar"})
```

Also, we should provide a description here.

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Format SQL in String Text Blocks";
+    }
+
+    @Override
+    public String getDescription() {
+        return "...";
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/sql/FormatSql.java:63]: A description would be nice here. For instance we should mention that this only applies to Java 17 text blocks.

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            @Override
+            public J.Literal visitLiteral(J.Literal lit, ExecutionContext ctx) {
+                J.Literal literal = super.visitLiteral(lit, ctx);
+                final SqlDetector detector = new SqlDetector();
+
+                if (isTextBlock(literal)) {
+                    String value = literal.getValue().toString();
+
+                    if (detector.isSql(value)) {
+                        String formatted = SqlFormatter.of(Dialect.valueOf(sqlDialect)).format(value);
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/sql/FormatSql.java]: The formatter being used here exposes some formatting options. Possibly we should expose the main options as options of the recipe. I am thinking of `indent`, `maxColumnLength`, and `uppercase`.

File: src/main/resources/META-INF/rewrite/sql.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+# https://www.apache.org/licenses/LICENSE-2.0
+# <p>
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+---
+type: specs.openrewrite.org/v1beta/recipe
```
Comment (knutwannheden) [inline src/main/resources/META-INF/rewrite/sql.yml:17]: For OpenRewrite this file isn't really necessary. What is the intended purpose of it?

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import com.github.vertical_blank.sqlformatter.SqlFormatter;
+import com.github.vertical_blank.sqlformatter.languages.Dialect;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class FormatSql extends Recipe {
+    @Option(displayName = "SQL dialect to be used to format SQL snippets.", description = "", example = "PostgreSql", required = false)
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/sql/FormatSql.java:44]: Description could use some love here.

File: src/main/resources/META-INF/rewrite/sql.yml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+name: org.openrewrite.sql.FormatTextBlocks
+displayName: Format SQL
+description: >
+  TODO
```
Comment (timtebeek) [inline src/main/resources/META-INF/rewrite/sql.yml:21]: I'm doubting if we need this file at all. It's not needed to run the recipe, as the Java class is also picked up already.

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                            FROM
+                              my_table
+                            WHERE
+                              something = 1;\\
+                            \""";
+                }
+                """
+            )
+            , 17)
+        );
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:62]: Unlikely edge case we might want to cover here is escaping existing `"""` in any query, and then cover that in the recipe as well.

File: build.gradle.kts [hunk 9]
```diff
@@ -12,6 +12,7 @@ dependencies {
     implementation("org.openrewrite:rewrite-core")
     implementation("org.openrewrite:rewrite-java")
     implementation("org.openrewrite:rewrite-yaml")
+    implementation("com.github.vertical-blank:sql-formatter:2.0.3")
```
Comment (knutwannheden) [inline build.gradle.kts:15]:
```suggestion
    implementation("com.github.vertical-blank:sql-formatter:2.0.+")
```

**Comment by timtebeek** [comment]
I've managed to fix the cycle issue as described by @shanman190 ; thanks for that!
I've tried `maybeAutoFormat`, but I suspect that does not cover text blocks, as all my attempts failed.

**Review by knutwannheden — COMMENTED** [review]
Thanks for the PR. I added a few review comments.

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java (hunk 8 above)
Comment (MahatmaFatalError) [reply src/test/java/org/openrewrite/sql/FormatSqlTest.java:62]: actually, there are many tests missing. I will try to add them later this week

File: src/main/java/org/openrewrite/sql/FormatSql.java (hunk 1 above)
Comment (MahatmaFatalError) [reply src/main/java/org/openrewrite/sql/FormatSql.java]: I'm not fully understanding your comment. Maybe the maintainers do and can commit the change.

File: src/main/java/org/openrewrite/sql/FormatSql.java (hunk 4 above)
Comment (MahatmaFatalError) [reply src/main/java/org/openrewrite/sql/FormatSql.java]: you mean the [FormatConfig](https://github.com/vertical-blank/sql-formatter/blob/master/src/main/java/com/github/vertical_blank/sqlformatter/core/FormatConfig.java) from https://github.com/vertical-blank/sql-formatter?search=1#examples?

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                J.Literal literal = super.visitLiteral(lit, ctx);
+                if (isTextBlock(literal)) {
+                    String originalValue = (String) literal.getValue();
+                    if (new SqlDetector().isSql(originalValue)) {
+                        String formatted = SqlFormatter.of(Dialect.valueOf(sqlDialect)).format(originalValue);
+                        TabsAndIndentsStyle style = getCursor().firstEnclosingOrThrow(SourceFile.class)
+                                .getStyle(TabsAndIndentsStyle.class);
+                        String indented = indent(literal.getValueSource(), formatted, style);
+                        return literal
+                                .withValue(indented)
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/sql/FormatSql.java:85]: The indentation should only be part of the `valueSource` not the `value`. This is a detail, but it could cause problems in other recipes if these two values don't match up.

```suggestion
                                .withValue(formatted)
```

Comment (timtebeek) [reply src/main/java/org/openrewrite/sql/FormatSql.java]: Yes I think I added that because I saw some cycle issues earlier; let's see if this broke the build and how to handle that then

File: src/main/java/org/openrewrite/sql/FormatSql.java (hunk 1 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/sql/FormatSql.java]: Applied in https://github.com/openrewrite/rewrite-sql/pull/2/commits/64b27fda1489affab52245f1bf023ab4e65a8f68

File: src/main/java/org/openrewrite/sql/SqlDetector.java [hunk 11]
```diff
@@ -128,6 +128,19 @@ private boolean probablySql(@Nullable String maybeSql) {
...
 
+    public boolean isSql(@Nullable String maybeSql) {
+        if (!probablySql(maybeSql)) {
+            return false;
+        }
+
+        try {
+            CCJSqlParserUtil.parse(maybeSql);
+        } catch (JSQLParserException e) {
+            return false; // not a valid SQL statement
```
Comment (MahatmaFatalError) [inline src/main/java/org/openrewrite/sql/SqlDetector.java:139]: should a log be added here?

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 12]
```diff
@@ -61,4 +61,324 @@ class Test {
...
+                              something = 1;\\
+                            \""";
+                }
+                """
+            )
+            , 15)
+        );
+    }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:95]: It's probably best to stick to one documentation example only
```suggestion
```

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 13]
```diff
@@ -61,4 +61,324 @@ class Test {
...
+                        );\\
+                      \""";
+                }
+              """
+            )
+            , 15)
+        );
+    }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:148]:
```suggestion
```

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 14]
```diff
@@ -61,4 +61,324 @@ class Test {
...
+                              (3, 'Star Lord', 2, 5);\\
+                            \""";
+                }
+                """
+            )
+            , 15)
+        );
+    }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:184]:
```suggestion
```

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 15]
```diff
@@ -61,4 +61,324 @@ class Test {
...
+                              1 OFFSET 1;\\
+                            \""";
+                }
+                """
+            )
+            , 15)
+        );
+    }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:317]:
```suggestion
```

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 16]
```diff
@@ -61,4 +61,324 @@ class Test {
...
+                              column1 is not null RETURNING *;\\
+                            \""";
+                }
+                """
+            )
+            , 15)
+        );
+    }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:354]:
```suggestion
```

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java [hunk 17]
```diff
@@ -61,4 +61,324 @@ class Test {
             , 15)
         );
     }
+
+    @DocumentExample
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/sql/FormatSqlTest.java:65]:
```suggestion
```

File: src/main/java/org/openrewrite/sql/FormatSql.java [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+import com.github.vertical_blank.sqlformatter.SqlFormatter;
+import com.github.vertical_blank.sqlformatter.languages.Dialect;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class FormatSql extends Recipe {
+    @Option(displayName = "SQL dialect to be used to format SQL snippets.", description = "Check out https://github.com/vertical-blank/sql-formatter#dialect for supported dialects.", valid = {"sql", "mariadb", "mysql", "postgresql", "db2", "plsql", "n1ql", "redshift", "spark", "tsql"}, example = "PostgreSql", required = false)
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/sql/FormatSql.java:44]: I'm guessing it's best if the example value matches on the valid values.
```suggestion
    @Option(displayName = "SQL dialect to be used to format SQL snippets.", description = "Check out https://github.com/vertical-blank/sql-formatter#dialect for supported dialects.", valid = {"sql", "mariadb", "mysql", "postgresql", "db2", "plsql", "n1ql", "redshift", "spark", "tsql"}, example = "postgresql", required = false)
```

File: src/main/java/org/openrewrite/sql/FormatSql.java (hunk 4 above)
Comment (MahatmaFatalError) [reply src/main/java/org/openrewrite/sql/FormatSql.java]: please re-check

File: src/test/java/org/openrewrite/sql/FormatSqlTest.java (hunk 8 above)
Comment (joanvr) [reply src/test/java/org/openrewrite/sql/FormatSqlTest.java:406]: Recipes in general shouldn't throw exceptions. For your example of an unknown dialect, which is restricted by the valid options, you could override the `validate` method to validate the input parameters of the recipe. You can see an example here: https://github.com/openrewrite/rewrite/blob/17b89bbd04dd6983918ddd5d2c81cdd954170347/rewrite-maven/src/main/java/org/openrewrite/maven/RemoveRedundantDependencyVersions.java#L80-L96
And some example tests to check the validation: https://github.com/openrewrite/rewrite/blob/17b89bbd04dd6983918ddd5d2c81cdd954170347/rewrite-maven/src/test/java/org/openrewrite/maven/AddManagedDependencyTest.java#L29-L52
You could also consider change the type of String to an enum, as you can see here: https://github.com/openrewrite/rewrite/blob/17b89bbd04dd6983918ddd5d2c81cdd954170347/rewrite-core/src/main/java/org/openrewrite/text/AppendToTextFile.java#L54-L64

Otherwise, the tests should check that no changes are made when it's not applicable, instead of checking for exceptions. In general, exceptions should be caught inside the recipe and return no results.

File: src/main/java/org/openrewrite/sql/SqlDetector.java [hunk 19]
```diff
@@ -128,6 +128,19 @@ private boolean probablySql(@Nullable String maybeSql) {
         return maybeSql != null && SIMPLE_HEURISTIC.matcher(maybeSql).find();
     }
 
+    public boolean isSql(@Nullable String maybeSql) {
+        if (!probablySql(maybeSql)) {
+            return false;
+        }
+
+        try {
+            CCJSqlParserUtil.parse(maybeSql);
```
[the remark this answers is not in the document]
Comment (MahatmaFatalError) [reply src/main/java/org/openrewrite/sql/SqlDetector.java:137]: Good point. As a side remark, maybe the heuristic is too coarse grained since it also rejects DDL which cis supported by the formatter

[the remark this answers is not in the document]
Comment (MahatmaFatalError) [reply src/main/java/org/openrewrite/sql/SqlDetector.java:144]: I changed the code that now CCJSqlParserUtil isn't involved at all. That brings also the benefit of supporting more SQL syntax but also the downside that much more false positives are formatted.

I would argue that limitation is OK, better too much than too less. The user can then decide to not commit certain lines in git.

**Comment by MahatmaFatalError** [comment]
I dont know what causes this compile problem
```
> Task :compileJava
Compiling with JDK Java compiler API.
/home/runner/work/rewrite-sql/rewrite-sql/src/main/java/org/openrewrite/sql/table/package-info.java:20: error: cannot find symbol
import org.openrewrite.PolyglotNamespace;
                      ^
  symbol:   class PolyglotNamespace
  location: package org.openrewrite
/home/runner/work/rewrite-sql/rewrite-sql/src/main/java/org/openrewrite/sql/package-info.java:20: error: cannot find symbol
import org.openrewrite.PolyglotNamespace;
                      ^
  symbol:   class PolyglotNamespace
  location: package org.openrewrite
/home/runner/work/rewrite-sql/rewrite-sql/src/main/java/org/openrewrite/sql/table/package-info.java:16: error: cannot find symbol
@PolyglotNamespace("OpenRewrite")
 ^
  symbol: class PolyglotNamespace
/home/runner/work/rewrite-sql/rewrite-sql/src/main/java/org/openrewrite/sql/package-info.java:16: error: cannot find symbol
@PolyglotNamespace("OpenRewrite")
 ^
  symbol: class PolyglotNamespace
4 errors
```

**Comment by timtebeek** [comment]
Your branch was two commits behind on `origin main` and needed some changes to adopt OpenRewrite 8.0; I've applied those changes just now.

File: src/main/java/org/openrewrite/sql/SqlDetector.java (hunk 11 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/sql/SqlDetector.java:147]: I think that might be a bit too common, as our heuristics are imperfect, and we have the catch clause to work around that. Any logged lines might then quickly become in-actionable noise.

File: src/main/java/org/openrewrite/sql/SqlDetector.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (timtebeek) [reply src/main/java/org/openrewrite/sql/SqlDetector.java]: I've reverted this change, as we incorrectly replaced some non-SQL statements without this duplicated check, which is in my opinion worse than checking it twice. I'd rather have a recipe that we can confidently run at scale rather than one where we have to monitor the output, and can not be run repeatedly because of that.

**Comment by timtebeek** [comment]
Thanks a lot for this recipe @MahatmaFatalError ! I've gone ahead and applied the final change I thought it needed to be able to safely run this broadly, and to separate some of the different functionality should we wish to revisit for instance the indentation later. All in all it's a very nice addition and I look forward to seeing it used in practice!

You should be able to try it out [using our snapshot versions](https://docs.openrewrite.org/reference/snapshot-instructions). Let me know how that works out for you! :)
